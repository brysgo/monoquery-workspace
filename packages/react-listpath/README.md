# React ListPath

[![CircleCI](https://circleci.com/gh/brysgo/react-listpath.svg?style=shield)](https://circleci.com/gh/brysgo/react-listpath)

_Let me start by saying I have no clue how this will be useful to you. For me, it is useful in the context of `react-monoquery` where you get graphql data encapsulation without having to pass down your data._

Say you have a component that sits in a couple nested lists. You don't feel like using any fancy data frameworks, and you don't want to have to pass indexes through react just to know what data to get from your component.

=> React List Path <=

It lets you wrap your lists with a `ListProvider` and then in your items, you can use the `ListPathConsumer` that will get back a render prop with an array of indexes.

## Usage

`yarn add react-listpath`

Then wrap the lists you want to track.

```jsx
const myData = [...];

const MyItem = ({someData}) => (<div>
  {someData}
</div>);

const MyPage = () => (
  myData.map((moreData) =>
    <div>{moreData.anotherList.map((someData) =>
      <MyItem someData={someData}/>
    )}</div>
 )
);
```

Could become:

```jsx
import { ListProvider, ListPathConsumer } from "react-listpath";

const myData = [];

const MyItem = ({ someData }) => (<ListPathConsumer>
  {listPath => <div>{myData[listPath[0]].anotherList[listPath[1]]}</div>}
</ListPathConsumer>);

const MyPage = () => (
  <ListProvider>
    myData.map((moreData) =>
    <div>
      <ListProvider>
        {moreData.anotherList.map((someData) =>
          <MyItem />
        )}
      </ListProvider>
    </div>
    )
 </ListProvider>
);
```

Why should I care about this? The example above doesn't give me any superpowers, it actually looks uglier the second way. However, in the context of `react-monoquery` I will not need to use the `ListPathConsumer`, since it will be queried for me, I just need to use the `ListProvider`, and then when I use the `Fragments` component my fragments will get me the right data magically without having to pass anything around. Fully declarative, single query composed react data without having to pass around props.