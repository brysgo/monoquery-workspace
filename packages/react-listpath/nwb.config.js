module.exports = {
  type: "react-component",
  npm: {
    esModules: true,
    umd: {
      global: "reactListPath",
      externals: {}
    }
  }
};
