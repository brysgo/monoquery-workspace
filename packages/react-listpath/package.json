{
  "name": "react-listpath",
  "version": "1.0.2",
  "description": "Track list item references through context",
  "repository": "git@github.com:brysgo/react-listpath.git",
  "author": "Bryan Goldstein <brysgo@gmail.com>",
  "main": "lib/index.js",
  "module": "es/index.js",
  "files": [
    "es",
    "lib",
    "umd"
  ],
  "scripts": {
    "build": "nwb build-react-component",
    "clean": "nwb clean-module",
    "test": "jest --config jest.config.js --no-cache",
    "prepublishOnly": "yarn build"
  },
  "devDependencies": {
    "babel-jest": "^23.2.0",
    "babel-plugin-transform-runtime": "^6.23.0",
    "babel-preset-env": "^1.6.1",
    "babel-preset-react": "^6.24.1",
    "jest": "^23.2.0",
    "nwb": "0.21.x",
    "react": "^16.4.2",
    "react-dom": "^16.3.2",
    "react-test-renderer": "^16.3.2"
  },
  "license": "MIT",
  "peerDependencies": {
    "react": "^16.3.2"
  }
}
