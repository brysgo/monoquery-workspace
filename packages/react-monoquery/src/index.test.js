import React from "react";
import graphql from "graphql-tag";

import { makeExecutableSchema } from "graphql-tools";
import { MockedProvider } from "react-apollo/test-utils";

import { MonoQuery, Fragments, gql } from "./";
import renderer from "react-test-renderer";

import { ListProvider } from "react-listpath";

const mockedData = {
  currentUser: {
    __typename: "User",
    name: "John Doe",
    email: "user@example.com",
    phone: "5555555555",
    orders: [
      {
        __typename: "Order",
        address: "1 North Pole",
        items: [
          {
            __typename: "Item",
            name: "only one thing"
          }
        ]
      },
      {
        __typename: "Order",
        address: "123 Street Street, Place, Country 12345",
        items: [
          {
            __typename: "Item",
            name: "Super cool stuff"
          },
          {
            __typename: "Item",
            name: "really cool nested thing"
          }
        ]
      }
    ]
  },
  otherThing: {
    __typename: "Thing",
    stuff: "foo"
  }
};

describe("when react-listpath is not installed", () => {
  beforeAll(() => {
    jest.mock("react-listpath", () => {
      throw new Error("Cannot find module 'react-listpath'");
    });
  });
  afterAll(() => {
    jest.unmock("react-listpath");
  });

  it("can build, run distribute simple query", async () => {
    const other = gql`
      fragment Other on User {
        email
      }
    `
    const fragments = {
      other,
      testing: gql`
        fragment TestingTesting on User {
          __typename
          name
          ...Other      
          phone
        }
        ${other}
      `
    };
    const query = gql`
      query MainQuery {
        currentUser {
          ...TestingTesting
        }
        otherThing {
          stuff
        }
      }
      ${fragments.testing}
    `;
    let done;
    const promise = new Promise(resolve => (done = resolve));
    const component = renderer.create(
      <MockedProvider
        mocks={[
          {
            request: {
              query: query.parsedQuery
            },
            result: { data: mockedData }
          }
        ]}
      >
        <MonoQuery query={query}>
          {data =>
            data.loading ? (
              <div>loading</div>
            ) : (
              (setTimeout(done, 0),
              (
                <div>
                  <Fragments fragments={fragments}>
                    {({ testing, other }) => (
                      <div>
                      <div>{JSON.stringify(testing, null, 2)}</div>
                      <div>{JSON.stringify(other, null, 2)}</div>
                      </div>
                    )}
                  </Fragments>
                </div>
              ))
            )
          }
        </MonoQuery>
      </MockedProvider>
    );
    await promise;
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe("with react-listpath", () => {
  it("supports nesting fragments in lists", async () => {
    const item = gql`
      fragment ItemFragment on Item {
        __typename
        name
      }
    `;
    const testing = gql`
      fragment AnotherTest on User {
        __typename
        name
        email
        phone
        orders {
          items {
            ...ItemFragment
          }
        }
      }
      ${item}
    `;
    const query = gql`
      query MainQuery {
        currentUser {
          ...AnotherTest
        }
        otherThing {
          stuff
        }
      }
      ${testing}
    `;
    let done;
    const promise = new Promise(resolve => (done = resolve));
    const component = renderer.create(
      <MockedProvider
        mocks={[
          {
            request: {
              query: query.parsedQuery
            },
            result: { data: mockedData }
          }
        ]}
      >
        <MonoQuery query={query}>
          {data =>
            data.loading ? (
              <div>loading</div>
            ) : (
              (setTimeout(done, 0),
              (
                <div>
                  <Fragments fragments={{ testing }}>
                    {({ testing }) => (
                      <div>
                        {JSON.stringify(testing, null, 2)}
                        <div>
                          <ListProvider>
                            {testing.orders.map(({ items }, i) => (
                              <ListProvider key={i}>
                                {items.map((x, j) => (
                                  <Fragments key={j} fragments={{ item }}>
                                    {({ item: { name } }) => <div>{name}</div>}
                                  </Fragments>
                                ))}
                              </ListProvider>
                            ))}
                          </ListProvider>
                        </div>
                      </div>
                    )}
                  </Fragments>
                </div>
              ))
            )
          }
        </MonoQuery>
      </MockedProvider>
    );
    await promise;
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
