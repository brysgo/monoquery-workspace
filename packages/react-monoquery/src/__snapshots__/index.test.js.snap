// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when react-listpath is not installed can build, run distribute simple query 1`] = `
<div>
  <div>
    <div>
      {
  "__typename": "User",
  "name": "John Doe",
  "phone": "5555555555"
}
    </div>
    <div>
      {
  "email": "user@example.com"
}
    </div>
  </div>
</div>
`;

exports[`with react-listpath supports nesting fragments in lists 1`] = `
<div>
  <div>
    {
  "__typename": "User",
  "name": "John Doe",
  "email": "user@example.com",
  "phone": "5555555555",
  "orders": [
    {
      "items": [
        {}
      ]
    },
    {
      "items": [
        {},
        {}
      ]
    }
  ]
}
    <div>
      <div>
        only one thing
      </div>
      <div>
        Super cool stuff
      </div>
      <div>
        really cool nested thing
      </div>
    </div>
  </div>
</div>
`;
